java/code/src/com/redhat/rhn/manager/contentmgmt/DependencyResolver.java

```diff
@@ -0,0 +1,178 @@
+        }
+
+        // Any other dependency filters (e.g. package dependencies) can be appended here
+        // TODO: This module can also be called at setup time to provide feedback using DependencyResolutionException
+
+        return updatedFilters;
+    }
+
+    /**
+     * Resolves modular dependencies and convert all module filters to package filters
+     *
+     * @param filters the list of module filters to be included in the project
+     * @return a list of package filters derived from the module filters, including dependencies
+     */
+    private List<PackageFilter> resolveModularDependencies(List<ModuleFilter> filters)
+            throws DependencyResolutionException {
+        List<Channel> sources = this.getSources();
+        List<Module> modules = filters.stream().map(ModuleFilter::getModule).collect(toList());
+        ModulePackagesResponse modPkgList;
+        try {
+            modPkgList = modulemdApi.getPackagesForModules(sources, modules);
+        }
+        catch (ConflictingStreamsException | ModuleNotFoundException e) {
+            throw new DependencyResolutionException("Failed to resolve modular dependencies.", e);
+        }
+
+        // 1. Modular packages to be denied
+        Stream<PackageFilter> pkgDenyFilters = modulemdApi.getAllPackages(sources).stream()
+                .map((String nevra) -> initFilterFromPackageNevra(nevra, ContentFilter.Rule.DENY));
+
+        // 2. Non-modular packages to be denied by name
+        Stream<PackageFilter> providedRpmApiFilters = modPkgList.getRpmApis().stream().map(
+                DependencyResolver::initFilterFromPackageName);
+
+        // 3. Modular packages to be allowed
+        Stream<PackageFilter> pkgAllowFilters = modPkgList.getRpmPackages().stream()
+                .map((String nevra) -> initFilterFromPackageNevra(nevra, ContentFilter.Rule.ALLOW));
+
+        // Concatenate filter streams into a list
+        return Stream.of(pkgDenyFilters, providedRpmApiFilters, pkgAllowFilters).flatMap(s -> s).collect(toList());
+    }
+
+
+    private static PackageFilter initFilterFromPackageNevra(String nevra, ContentFilter.Rule rule) {
+        FilterCriteria criteria = new FilterCriteria(FilterCriteria.Matcher.EQUALS, "nevra", nevra);
+        PackageFilter filter = new PackageFilter();
+        filter.setRule(rule);
+        filter.setCriteria(criteria);
+        return filter;
+    }
+
+    private static PackageFilter initFilterFromPackageName(String name) {
+        FilterCriteria criteria = new FilterCriteria(FilterCriteria.Matcher.MATCHES, "name", name);
+        PackageFilter filter = new PackageFilter();
+        filter.setRule(ContentFilter.Rule.DENY);
+        filter.setCriteria(criteria);
+        return filter;
+    }
+
+    /**
+     * Get a list of modular channels among the sources
+     * @return the list of modular channels
+     */
+    private List<Channel> getSources() {
+        return project.getSources().stream()
```

Comment:
I'm not sure getting **all** sources of a project in this situation is the right thing. I think we should filter out `DETACHED` sources, i.e. the sources flagged for removal after the project build. We had a similar situation in past for filters (they also support "soft-delete" mechanism) and for this purpose, we implemented a helper method `ContentProject.getActiveFilters`. Maybe we need something similar for the sources.

Reply:
That's a good point. I added `ContentProject.getActiveSources`, filtering out the detached ones.